Issue 892

RyanDJLee opened the pull request:
Create module containing functions and data types to create and display Course Requirements and Program Requirements.

david-yz-liu commented on app/Database/Requirement.hs:
```diff
@@ -0,0 +1,59 @@
+-- | Returns a well formatted String representing a program requirement for specified program.
+showProgramReq :: ProgramReq -> String
+showProgramReq (PRGREQ program reqs) = "The program requirements for " + program + ":\n"
+										+ map (showReq) reqs
```
You don't need parentheses around `showReq`

david-yz-liu replied:
(This seems to occur in multiple places)

david-yz-liu commented on app/Database/Requirement.hs:
```diff
@@ -0,0 +1,59 @@
+										+ map (showReq) reqs
+
+
+-- Assuming we have correctly parsed inputs, CourseReq is in order [corequisite, exclusion, prerequisite]
```
This isn't a good comment - you shouldn't assume anything about the order of the requirements in the list `[CourseReq]`.

david-yz-liu commented on app/Database/Requirement.hs:
```diff
@@ -0,0 +1,59 @@
+
+
+-- for now J seems to be most readable and convenient value constructor for satisfying rec structure.
+data Req = J String | AND [Req]| OR [Req] | FROM (Integer) [Req]
```
Missing a space before the second `|`. Don't need parentheses around `(Integer)`.

david-yz-liu commented on app/Database/Requirement.hs:
```diff
@@ -0,0 +1,59 @@
+data CourseRequirement = CRSREQ String [CourseReq]
+
+-- | Returns a well formatted String representing all course requirements for specified course.
+showCourseRequirement :: CourseRequirement -> String
```
Read through [here](http://learnyouahaskell.com/making-our-own-types-and-typeclasses#typeclasses-102) to learn how to make `CourseRequirement`, and `Req` instances of the `Show` typeclass. (There's only one function you need to implement, `show`.)

david-yz-liu commented on app/Database/Requirement.hs:
```diff
@@ -0,0 +1,59 @@
+showCourseRequirement (CRSREQ course coursereqs) = course + "\n" + map (showCourseReq) coursereqs 
+
+
+data CourseReq = CREQ String Req | EXCL String Req | PREQ String Req
```
You don't need this type. `CourseRequirement` can have type `CRSREQ String Req Req Req`, where the first `Req` is the corequisite, the second is the exclusion, and the third is the prerequisite.

RyanDJLee commented:
Removed unnecessary parentheses and fixed syntax and spacing errors.

Changed data type contract for CourseReq type and removed CourseRequisite type as it was no longer necessary.

Made all data types instances of Show class.

david-yz-liu commented on app/Database/Requirement.hs:
```diff
@@ -0,0 +1,39 @@
+
+-- | Returns a well formatted String representing a program requirement for specified program.
+instance Show ProgramReq where
+	show (PRGREQ program reqs) = "Program Requirements for " ++ program ++ ":/n" ++ map show reqs
```
I'm not sure if the `/n` is supposed to be a newline character (`\n`)?

david-yz-liu commented on app/Database/Requirement.hs:
```diff
@@ -0,0 +1,39 @@
+
+instance Show Req where
+	show (J course) = course
+	show (AND x:xs) = show x ++ ", " ++ show AND xs
```
`show AND xs` should be `show (AND xs)`, but even better to use a `map` plus the cool Haskell function `intercalate`

RyanDJLee commented:
Fixed syntax for newline character.

Implemented show function using intercalate and map.

david-yz-liu commented on app/Database/Requirement.hs:
```diff
@@ -0,0 +1,39 @@
+
+-- | Returns a well formatted String representing a program requirement for specified program.
+instance Show ProgramReq where
+	show (PRGREQ program reqs) = "Program Requirements for " ++ program ++ ":\n" ++ map show reqs
```
It seems you're using tabs instead of spaces; make sure to use exactly 4 spaces for indentation.

david-yz-liu commented on app/Database/Requirement.hs:
```diff
@@ -0,0 +1,39 @@
+
+instance Show Req where
+	show (J course) = course
+	show (AND x:xs) = S.intercalate "," $ map show [x, (AND xs)]
```
You don't need to break up the arguments to `AND`. Do `show (AND reqs) = ... map show reqs`

RyanDJLee commented:
- Converted indents to 4 spaces.

- Implemented Show Req using better method.
